src/lib.rs

```diff
@@ -0,0 +1,512 @@
+#[derive(thiserror::Error, PartialEq)]
+pub enum Error {
+    #[error("configuration error: {0}")]
+    ConfigError(String),
+    #[error("API error: {0}")]
+    ApiError(String),
+    #[error("callback error: {0}")]
+    CallbackError(String),
+    #[error("feature not implemented: {0}")]
+    NotImplementedError(String),
+}
+
+impl std::fmt::Debug for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Error::NotImplementedError(e)
+            | Error::ConfigError(e)
+            | Error::ApiError(e)
+            | Error::CallbackError(e) => {
+                write!(f, "{}", e)
+            }
+        }
+    }
+}
+
+/// EvidenceBuilder is the function signature of the application callback.
+/// The application is passed the nonce and the list of supported evidence media
+/// types and shall returns the computed evidence together with the selected
+/// media type.
+type EvidenceBuilder =
+    fn(nonce: &Vec<u8>, accepted: &Vec<String>) -> Result<(Vec<u8>, String), Error>;
+
+pub struct ChallengeResponseConfig {
+    nonce: Option<Vec<u8>>,
+    nonce_sz: usize,
```

Comment:
the trouble here is that when the nonce size is supplied, it's veraison that will create the nonce and make it available through the session resource, not the library.

Reply:
I think I'd need to understand this better, but it'd probably make sense to have it as an enum anyway, just to prevent inconsistencies (via typing). Though I might be misunderstanding something more fundamental!

Reply:
oh, sorry for misunderstanding.  I didn't know Rust enum's where that powerful!  That sounds like an excellent idea indeed.

Reply:
done in [660ba8e](https://github.com/veraison/rust-apiclient/pull/4/commits/660ba8e4f56e0dccc864264c60cd235810adc454)